### PR TITLE
ceph: remove atomic_t

### PR DESCRIPTION
**Pull request by chardan**
This is intended to remove atomic_t from the codebase, replacing uses of
it with the C++ standard library. It should add no features and cause no
regressions. The "include/atomic.h" header file and "common/simple_spin.h"
are removed.

Please be sure to inspect default values that I've assigned to atomics-- I
occasionally guessed-- and check for any undefined behavior I may have
introduced. I'm also not entirely sure if the removal of InstanceReplayer.h
is something I may have done through misuse of git, or something expected.
Guidance appreciated.

I've taken a pretty mechanical approach here-- for example, replacing
read() with load(). I wouldn't mind using std::atomic<>'s type coercion for a 
more natural syntax  (x = y as opposed to x = y.load()).

Another refinement might be to apply a template to help with situations 
where we compare-and-swap, but intended to discard the swapped-to value (
avoiding having to add exposed variables for compare_exchange_*).

Please let me know if you'd like to see either or both of these included
as additional commits.

Thank you!

### CODE REVIEW

**Comment by chardan** [comment]
@cbodley @smithfarm @liewegas

File: src/include/Spinlock.h [hunk 1]
```diff
@@ -16,88 +16,49 @@
...
-  return pthread_spin_init(&l->lock, PTHREAD_PROCESS_PRIVATE);
+ while(lock.test_and_set(std::memory_order_acquire))
+  ;
 }
 
-static inline int ceph_spin_destroy(ceph_spinlock_t *l)
+inline void simple_spin_unlock(std::atomic_flag& lock)
 {
-  return pthread_spin_destroy(&l->lock);
+ lock.clear(std::memory_order_release);
```
Comment (tchaikov) [inline src/include/Spinlock.h]: again, do we have a micro benchmark result to compare the performance with and without the change?

Comment (chardan) [inline src/include/Spinlock.h]: Not yet. I'll see what I can do.

**Comment by liewegas** [comment]
Hmm, like other cleanups, this will be easier if we break it up into a few different commits and PRs targetting different subsystems.  They'll have to go through different teuthology suites and possibly get rebased to avoid conflicts with other PRs.  Can you break this into osd/, mds/, librbd/, rgw/, and then everything else (common)?  The last patch that removes the old atomic.h will come separately and last.

File: src/include/Spinlock.h (hunk 1 above)
Comment (chardan) [inline src/include/Spinlock.h]: It's quick-n-dirty, but here's a start. Please feel free to kick the tires:
https://github.com/chardan/ceph-spinlock-minibench

**Comment by chardan** [comment]
@liewegas If there are too many separate commits, I can also squish some of them together.

**Comment by liewegas** [comment]
@chardan commits are fine. Can you break it into a few separate PRs though so that one conflict in say rbd doesn't block everything else from test/merge?  Thanks!

**Comment by chardan** [comment]
Further adventures continue at:
https://github.com/ceph/ceph/pull/14655

**Pull request closed without merging**